### ISSUE TITLE
Change scopeType matchers to rely on tree-sitter style scheme queries

### ISSUE THREAD

**Issue by Will-Sommers** [cursorless-dev/cursorless#616]
### Background
Currently Cursorless uses a custom pattern definition DSL alongside a set of helper functions in [nodeMatcher.ts](https://github.com/cursorless-dev/cursorless/blob/6c63c210a217549cebd4ff846f4eaff9c23d50f7/src/util/nodeMatchers.ts) to match various scope types, such as `item` within a list or `argue` within a function definition or function invocation. 

The tree-sitter project also provides a DSL, written in Scheme which allows a user to query for patterns within syntax trees. Here's a [link to the docs](https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries) and an [example usage in JS](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/test/query-test.js) via the web-tree-sitter project. Each query then is allowed to assign a `name` to a node, such as [`@comment`](https://github.com/tree-sitter/tree-sitter-ruby/blob/master/queries/highlights.scm#L133-L137) or [`@punctuation.bracket`](https://github.com/tree-sitter/tree-sitter-ruby/blob/master/queries/highlights.scm#L133-L137):

```
(comment) @comment

[
  "("
  ")"
  "["
  "]"
  "{"
  "}"
  "%w("
  "%i("
] @punctuation.bracket
```

The name can then be read or [asserted against](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/test/query-test.js#L63). 

The thought is that moving towards this approach will be more expressive out of the box. Additionally, many other projects including Neovim and Helix rely on queries for syntax highlighting as well as indentation which might help to make the incremental work for adding a new language a little bit simpler, since there are already partial or full definitions to work from.  In particular, Helix already uses these queries for their [textobjects](https://docs.helix-editor.com/usage.html#textobjects), which is a simplified version of Cursorless scope types.  Here's [an example](https://github.com/helix-editor/helix/blob/master/runtime/queries/python/textobjects.scm) of a set of textobject definitions; they exist for several other languages as well

### The Work
- Create a `queries` directory with a subdirectory for each language, eg `queries/python`, etc
- Create a query file for a language that provides queries for some or all of these [`ScopeTypes`](https://github.com/cursorless-dev/cursorless/blob/main/src/typings/Types.ts#L96-L127), placing the file in `queries/<language>/scopeTypes.scm`
- Add the ability to load queries on a per-language basis
- Queries occur on the tree level (`SyntaxNode.Tree.Language`) and so are top down rather than bottom up as cursorless node matchers currently work. 
- Should a query have successful matches, return the smallest range containing the input selection
- Note that for some of the auxiliary definitions listed below (eg `@<scopeType>.searchScope`) we first find a match, and then [search within that range](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/test/query-test.js#L74-L75)

### The definitions

- Default query tag is just `@<scopeType>`, so eg `@namedFunction`
- In addition, we support a few other queries:
  - `@<scopeType>.removalRange` indicates a different range that should be used for removal
  - `@<scopeType>.domain` indicates that we should first expand to the smallest containing match for this tag and then search for a rooted instance of `@<scopeType>` within this region.  The canonical example for this one is enabling `take value` from within the key in a map: we'd set `@collectionItem.domain` to be the containing pair
  - `@<scopeType>.iterationScope` indicates that when user says `"every <scopeType>"`, we should first expand to the smallest instance of this tag, and then yield all top-level instances of `@<scopeType>` within this range.  Here, top-level means not contained by any other match within the search range.  Also, note that when finding the instances in the range, we should use `@<scopeType>.domain` if it exists.  See below for an explanation
  - `@<scopeType>.interior` is used by `excludeInterior` and `interiorOnly` stages (see #254)

### Migration notes

This will require a replacement of each of the language matcher files with a `scopeTypes.scm` definition. For this reason, we will want to support both paths while the migration occurs.  We can keep doing continuous delivery during migration because every language other than C# is well tested. 

### Questions

- Do we want to change our term `scopeType` to `textObject`?  That is the term used in both nvim tree-sitter, helix, and by [redstart voice](http://redstartsystems.com/human-machine-grammar-dictionary)
- Better term for `@<scopeType>.iterationScope`?
  - `@<scopeType>.parent`?
- How to handle argument lists and collection items?  I have a feeling we'll be repeating `,` stuff for removal ranges a lot.  I wonder if we want to add Toml configuration for languages where we can indicate scopes that should be handled as comma-separated lists.  Along this direction, it's worth thinking about the connection to #357 
- Do we want to support custom queries?  [Here's](https://github.com/nvim-treesitter/nvim-treesitter#adding-queries) how neovim does it
- Do we still want to support the "every" in cases where a scope doesn't explicitly specify a `@<scopeType>.domain`?  We do that today by just iterating the parent.  Might be useful to keep this one as a fallback 🤷‍♂️

#### Why we need to use `@<scopeType>.domain` when searching within `@<scopeType>.iterationScope`

Consider the following case:

```typescript
{
   foo: {
      bar: "baz"
   }
}
```

If the user says `"take every key fine"`, we want to just return `foo`, excluding the nested key `bar`.  In this case `key.iterationScope` is `object` and `key.domain` is `pair`.  If we just looked for instances of `key` within the `object`, we'd get the nested key as well.  However, if we search for top-level `pair` objects we won't, as desired

#### Why `@<scopeType>` must be rooted within `@<scopeType>.domain`

We can actually use the same code example as above:

```typescript
{
   foo: {|
      bar: "baz"
   }
}
```

If the user says "take key" with the cursor at the indicated position (after second opening bracket), we want to select `foo`.  We first expand to the containing `pair`, as that is the definition of `key.domain`.  Then we need to find the `key`.  If we just look for top-level `key`s (ie not contained by other `key`s), we'll end up with both `foo` and `bar`.  If we require that the `key` be rooted within the `pair`, that won't happen

Fwiw, we could possibly instead exclude any `@<scopeType>` matches which are contained within a lower `@<scopeType>.domain`

### Resources

- Helix [documentation](https://docs.helix-editor.com/guides/textobject.html)
- Helix [implementation](https://github.com/helix-editor/helix/blob/3e78b8fdad953a14c9a4f815295dfc6ed75d409f/helix-core/src/textobject.rs#L147-L151)
- Helix [queries](https://github.com/helix-editor/helix/tree/master/runtime/queries)
- nvim-treesitter [documentation](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#parser-configurations)
- nvim-treesitter implementation (might actually be in neovim itself)
- nvim-treesitter [queries](https://github.com/nvim-treesitter/nvim-treesitter/tree/master/queries)
- neovim [documentation](https://github.com/neovim/neovim/blob/master/runtime/doc/treesitter.txt)
- neovim implementation
  - https://github.com/neovim/neovim/blob/master/runtime/lua/vim/treesitter.lua
  - https://github.com/neovim/neovim/tree/master/runtime/lua/vim/treesitter
- neovim [tests](https://github.com/neovim/neovim/tree/master/test/functional/treesitter)

### Addenda:
- The query syntax also allows for the [injection of information](https://github.com/tree-sitter/tree-sitter-html/blob/master/queries/injections.scm#L3) apart from the `name` of a node and using this DSL will likely be the approach used to support [multi-language documents](https://github.com/cursorless-dev/cursorless-vscode/issues/409).
- https://github.com/cursorless-dev/cursorless/issues/685

### Attributions

👋 Big H/T 🎩 to @wenkokke for the original idea

**Comment by wenkokke** [comment]
> `@<scopeType>.removalRange` indicates a different range that should be used for removal

Using things like `.removalRange` has as a disadvantage that we can only include a node in *one such range*.

**Comment by wenkokke** [comment]
> `@<scopeType>.searchScope` indicates that we should first expand to the smallest containing match for this tag and then search for `@<scopeType>` within this region. The canonical example for this one is enabling take value from within the key in a map: we'd set `@collectionItem.searchScope` to be the containing pair

My proposal featured a simplified query language for delimiters, which we search upwards from the current token. It would be good to somehow merge these proposals, to have a concept like “all results for query Q2 within the node returned by Q1 containing the mark”.

**Comment by wenkokke** [comment]
> `@<scopeType>.iterationScope` indicates that when user says "every <scopeType>", we should first expand to the smallest instance of this tag, and then yield all top-level instances of `@<scopeType>` within this range.

I propose we do a slightly different thing, using the delimiter mechanism described above. If we run Q2 within the delimiting scope, then “every” simply means “all matches” and “<ordinal>” means the `ordinal`’th match. The default would be to again return the node which contains the mark.

**Comment by pokey** [comment]
> > `@<scopeType>.removalRange` indicates a different range that should be used for removal
> 
> Using things like `.removalRange` has as a disadvantage that we can only include a node in _one such range_.

Sorry not sure I follow.  Can you elaborate on the problem?

**Comment by wenkokke** [comment]
> > query language for delimiters
> 
> Can you elaborate on what you mean by this? What kind of delimiters?

For, e.g., function parameters in a definition in Haskell, I search upwards to the first function definition node, which will delimit the second query, a downwards search for parameters rooted at that definition.

I think this achieves something similar to @Will-Sommers suggestion of using an `.iteratorRange` annotation, but theirs is cleaner, since it manages to use a single query and a single query language.

**Comment by pokey** [comment]
> > > query language for delimiters
> > 
> > 
> > Can you elaborate on what you mean by this? What kind of delimiters?
> 
> For, e.g., function parameters in a definition in Haskell, I search upwards to the first function definition node, which will delimit the second query, a downwards search for parameters rooted at that definition.
> 
> I think this achieves something similar to @Will-Sommers suggestion of using an `.iterationScope` annotation, but theirs is cleaner, since it manages to use a single query and a single query language.
> 
> Theirs would search the entire document for all function definitions with parameters, then grab the one which contains the mark, and then (probably could) return a list of captures corresponding to the parameters, either taking every, the n’th, or by default the one containing the mark.
> 
> Would even, I just realised, open up the suggestion of saying things like “take next arg” to take the parameter _one after_ the one which contains the cursor, and other such relative commands.

I'm bit confused: are you just suggesting that we just do what's described in the issue, or are you proposing that we change something?

**Comment by pokey** [comment]
@Will-Sommers maybe tomorrow let's go through the exercise of writing a language definition?  I think we might be able to make some headway in an hour and it should help to make things more concrete

**Comment by wenkokke** [comment]
> Ok I'm confused: are you suggesting that we just do what's described in the issue, or are you proposing that we change something?

I propose we use @Will-Sommers approach of marking — what I call — delimiters using a named capture, then select all matches where the delimiter is the smallest possible region containing the mark, and use my approach to assign captures to every/ordinal/default.

**Comment by pokey** [comment]
Fwiw here is @wenkokke's original proposal:

[Cursorless NodeFinder.pdf](https://github.com/cursorless-dev/cursorless/files/8336751/Cursorless.NodeFinder.pdf)

**Comment by pokey** [comment]
Fwiw the reason that I was leaning away from rooted queries and towards top-level matches was so that we could support "take every key" both with nothing selected, which would expand to containing map, and with something selected, which would search within selection.  If we use top-level matches, then we can handle both in the same way

**Comment by pokey** [comment]
Ok updated issue description to clarify iteration implementation a bit.  See last revision of description to see the change

<img width="648" alt="history" src="https://user-images.githubusercontent.com/755842/159804615-9a26b5d4-b3b7-4658-9307-077172c3a628.png">

**Comment by pokey** [comment]
Ok I captured our mob programming session in draft PR #620

**Comment by wenkokke** [comment]
> Fwiw the reason that I was leaning away from rooted queries and towards top-level matches was so that we could support "take every key" both with nothing selected, which would expand to containing map, and with something selected, which would search within selection. If we use top-level matches, then we can handle both in the same way

What about something like “take every key in funk red made”, where we have “in” as a composing keyword to do the second search (every key) in the scope of the first (funk red made).

**Comment by pokey** [comment]
Yes, that's the idea (see [here](https://github.com/cursorless-dev/cursorless/wiki/Target-overhaul), [here](https://github.com/cursorless-dev/cursorless/wiki/What's-next#programming-language), and [here](https://github.com/cursorless-dev/cursorless/issues/69)), but the issue is that you could end up with subtly different behaviour if we're not careful, because that style of command is unrooted, because the parent doesn't come from a query

**Comment by pokey** [comment]
Also @wenkokke can we align on naming here?  I'm using `@<scope_type>.searchScope` instead of your `@<scope_type>.delimiter`.  When I hear the term "delimiter", I assume something like a `(` or `,`.  To me `searchScope` is clearer.  Absolutely open to discussing terms, but let's try to at least align to avoid confusion 😅

**Comment by wenkokke** [comment]
> Also @wenkokke can we align on naming here? I'm using `@<scope_type>.searchScope` instead of your `@<scope_type>.delimiter`. When I hear the term "delimiter", I assume something like a `(` or `,`. To me `searchScope` is clearer. Absolutely open to discussing terms, but let's try to at least align to avoid confusion 😅

I’m using delimiter in the sense of (parse) trees or e.g. delimited continuations, i.e., a node which delimits a subtree, but iteratorScope is probably more generally clear? I’d prefer to use something like “punctuation” for what you call delimiters, but I’m fine either way.

**Comment by wenkokke** [comment]
I’m using “delimiter” the way @Will-Sommers uses “iteratorScope”, I think? I don’t fully understand “searchScope”?

**Comment by Will-Sommers** [comment]
Just a small note — looks like nvim has added some stuff here as well: 
https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/master/queries/javascript/textobjects.scm#L81

**Comment by pokey** [comment]
Ooh nice find. Interesting to see how they've handled ranges

**Comment by Will-Sommers** [comment]
https://github.com/nvim-treesitter/nvim-treesitter/blob/a838a35b2b5514cc341edfe398dc5036df554ac4/lua/nvim-treesitter/query.lua#L233-L246

**Comment by Will-Sommers** [comment]
Heyo, so it does look like user supported. What ends up happening is that a `textPredicate` is registered an returned on each query. An operate and then a set of operands is set and it looks like we can match either off of a [capture or a string match](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/test/query-test.js#L291-L296). The `operand` would dictate the behavior which is implemented on the TS side. 

From that point, the ball is back in our court([see above link](https://github.com/nvim-treesitter/nvim-treesitter/blob/a838a35b2b5514cc341edfe398dc5036df554ac4/lua/nvim-treesitter/query.lua#L237)). 

Going to play around with implenting `#make-range!` and update the draft PR.

**Comment by pokey** [comment]
Yep sounds good to me. I think ideally we'll just need to implement a few simple generic custom functions

**Comment by Will-Sommers** [comment]
Heyo @pokey —

Just to catch you up on my thoughts. The PR I'm working on will not define `searchScope`, nor will it implement any matchers which would require a search scope, eg. argue, item, key. Instead, I'm going to keep the current scopes in the file(see below). I've implemented the default "parent" matching, relying on it only if a `searchScope` isn't defined. I'll clean this up and get a PR up.

All of this will happen within the returned matcher function and we'll implement it there, adding `  siblings?: boolean` to `NodeValueMatcher`. I think this will be a lot cleaner than relying non-encapsulated behavior within `getNodeMatcher`. 

I'm working on some tests for the code itself now. I'm not sure the philosophy related to unit tests(most everything looks like functional integration tests rather than unit.


```
(comment) @comment
(method) @function
(if) @if
(call) @call
(method) @namedFunction
(method
  name: (_) @functionName)
(hash) @map

[
  (array)
  (string_array)
  (symbol_array)
] @list

(regex) @regex

(class) @class
(class
  name: (_) @className)

(assignment
  left: (_) @name)
(operator_assignment
  left: (_) @name)
(class
  name: (_) @name)
(method
  name: (_) @name)
```

**Comment by pokey** [comment]
Just chatted with @AndreasArvidsson and we came up with a nice name for `searchScope`: `enclosure`.  I think that captures the idea much more clearly.  So for example, in the query

```query
(operator_assignment
  left: (_) @name) @name.enclosure
```

Then internally the object that describes what came back would have `enclosureRange` and `extractionRange`, the latter of which would be `@name` above

**Comment by Will-Sommers** [comment]
@pokey @AndreasArvidsson — this works for me but I'll likely bring it into another PR after the initial one was merged. 

Here's the [diff of my work on it](https://github.com/Will-Sommers/cursorless-vscode/compare/scheme-scope-demo...Will-Sommers:search-scope?expand=1), the `@name.enclosure` is flipped in my version but I think you're right in having it this way. It sounds like I might need to read up on the compositional modifiers PR to see how this will tie in

**Comment by pokey** [comment]
other possible terms for `enclosure`:

- `dominion`
- `domain`

the notion is that we want to indicate the range within which this scope is the canonical instance of the given scope type.  So eg for `key`, its domain is `pair`: within a given pair it is the canonical key

**Comment by wenkokke** [comment]
> Just chatted with @AndreasArvidsson and we came up with a nice name for `searchScope`: `enclosure`. I think that captures the idea much more clearly. So for example, in the query
> 
> ```
> (operator_assignment
>   left: (_) @name) @name.enclosure
> ```
> 
> Then internally the object that describes what came back would have `enclosureRange` and `extractionRange`, the latter of which would be `@name` above

Why not just `@name.scope`?

**Comment by wenkokke** [comment]
Otherwise, what is currently preventing this pull request from being merged?

**Comment by pokey** [comment]
`scope` is a bit too generic; we ended up aligning on `domain`, so glad to hear you like it as well.  Fwiw we're doing some pairing on this one; happy to loop you in if interested

The PR needs some various fixes; I think @Will-Sommers has a todo list based on our most recent pairing session

**Comment by pokey** [comment]
> Dominion is reserved for `@🐯` or `@simba`. Using enclosure with `@🐯` or `@simba` is animal cruelty.

😄

**Comment by AndreasArvidsson** [comment]
We are today relying on Tree sitter queries and most languages are migrated
